tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/field_access.bal

```diff
@@ -353,10 +353,84 @@ function testFieldAccessOnMapConstruct() returns boolean {
     return "Sanjiva" == name;
 }
 
+type FooRec1 record {
+    json j;
+};
+
+type FooRec2 record {
+    json j;
+    FooRec1 fooRec;
+};
+
+function testFieldAccessOnJsonTypedRecordFields() returns error? {
+    FooRec1 rec1 = {j: "1"};
+    string val = check rec1.j;
+    assertEquals(val, "1");
+
+    rec1 = {j: {k: "2"}};
+    val = check rec1.j.k;
+    assertEquals(val, "2");
+
+    rec1 = {j: {k: {l: "1", m: "3"}}};
+    val = check rec1.j.k.m;
+    assertEquals(val, "3");
+
+    FooRec2 rec2 = {j: "1", fooRec: {j: "2"}};
+    val = check rec2.j;
+    assertEquals(val, "1");
+    val = check rec2.fooRec.j;
+    assertEquals(val, "2");
+
+    rec2 = {j: "1", fooRec: {j: {k: "3", l: {m: "4"}}}};
+    val = check rec2.fooRec.j.k;
+    assertEquals(val, "3");
+    val = check rec2.fooRec.j.l.m;
+    assertEquals(val, "4");
+}
+
+function testFieldAccessOnJsonTypedRecordFieldsResultingInErrors() {
+    FooRec1 rec1 = {j: "1"};
+    json|error val = rec1.j.k.l;
```

Comment:
Shall we add a test where the field is not present also? E.g., `k` is not a key in `rec1.j`.

Reply:
Are you referring to this case `rec1.a`?

Reply:
Yes. And `rec1.a.b`.

Reply:
For this case we get this compile-time error `invalid field access: 'a' is not a required field in record 'FooRec1', use member access to access a field that may have been specified as a rest field` on 2.x and 3.0 . Seems not related to the issue. Do we need to add that case?

Reply:
Sorry, meant on `rec1.j`. `rec1.j.a`, `rec1.j.a.b`

Reply:
Updated 8ce9bb79f0adc9cf67ab12d17decf545ba190bdc

Reply:
Shall we add one or two tests with optional field access also?

E.g.,
```ballerina
type Rec record {
    json j?;
};

public function main() returns error? {
    Rec r = {};
    string s = check r?.j.a.b;
    ...
}
```

Reply:
Updated 701732def6e865ec2220e546cf31ec9f60b825a6. 
But found a bug in master 
```
import ballerina/io;

type FooRec1 record {
    json j;
};

type FooRec3 record {
    json j?;
    FooRec1 fooRec?;
};

public function main() returns error? {
    FooRec3 rec2 = {j: "1", fooRec: {j: {k: "3", l: {m: "4"}}}};
    json val = check rec2.fooRec?.j.k;
    io:println(val); // got {"k":"3","l":{"m":"4"}} which is wrong
}
```
The value for `val` is wrong.